swarm/storage/resource.go

```diff
@@ -0,0 +1,528 @@
+*
+* the data itself is prefixed with a signed hash of the data. The sigining key
+* can be used to verify the authenticity of the update, for example by looking
+* up the ownership of the namehash in ENS and comparing to the address derived
+* from it
+*
+* NOTE: the following is yet to be implemented
+* The resource update chunks will be stored in the swarm, but receive special
+* treatment as their keys do not validate as hashes of their data. They are also
+* stored using a separate store, and forwarding/syncing protocols carry per-chunk
+* flags to tell whether the chunk can be validated or not; if not it is to be
+* treated as a resource update chunk.
+*
+* TODO: signature validation
+ */
+
+const (
+	SIGNATURE_LENGTH = 65
+)
+
+// Encapsulates an actual resource update. When synced it contains the most recent
+// version of the resource update data.
+type resource struct {
+	name       string
+	ensname    common.Hash
+	startblock uint64
+	lastblock  uint64
+	frequency  uint64
+	version    uint64
+	data       []byte
+	updated    time.Time
+}
+
+// Main interface to resource updates. Creates/opens the resource database,
+// and sets up rpc client to the blockchain api
+type ResourceHandler struct {
+	ChunkStore
+	ethapi       *rpc.Client
+	resources    map[string]*resource
+	hashLock     *sync.Mutex
+	resourceLock *sync.Mutex
+	hasher       SwarmHash
+	privKey      *ecdsa.PrivateKey
+}
+
+// Create or open resource update chunk store
+func NewResourceHandler(privKey *ecdsa.PrivateKey, datadir string, cloudStore CloudStore, ethapi *rpc.Client) (*ResourceHandler, error) {
+	path := filepath.Join(datadir, "resource")
+	dbStore, err := NewDbStore(datadir, nil, singletonSwarmDbCapacity, 0)
+	if err != nil {
+		return nil, err
+	}
+	localStore := &LocalStore{
+		memStore: NewMemStore(dbStore, singletonSwarmDbCapacity),
+		DbStore:  dbStore,
+	}
+	hasher := MakeHashFunc("SHA256")
+	return &ResourceHandler{
+		ChunkStore:   newResourceChunkStore(path, hasher, localStore, cloudStore),
+		ethapi:       ethapi,
+		resources:    make(map[string]*resource),
+		resourceLock: &sync.Mutex{},
+		hashLock:     &sync.Mutex{},
+		hasher:       hasher(),
+		privKey:      privKey,
+	}, nil
+}
+
+// Creates a standalone resource object
+//
+// Can be passed to SetResource if external root data lookups are used
+func NewResource(name string, startblock uint64, frequency uint64) (*resource, error) {
+	validname, err := idna.ToASCII(name)
+	if err != nil {
+		return nil, err
+	}
+
+	return &resource{
+		name:       validname,
+		ensname:    ens.EnsNode(validname),
+		startblock: startblock,
+		frequency:  frequency,
+	}, nil
+}
+
+// Creates a new root entry for a resource update identified by `name` with the specified `frequency`.
+//
+// The start block of the resource update will be the actual current block height of the connected network.
+func (self *ResourceHandler) NewResource(name string, frequency uint64) (*resource, error) {
```

Comment:
I can see that we have validation of `frequency` in `SetResource` (check that it is not 0). Shouldn't we have validation for `frequency` in this method as well?

Reply:
I've removed the check in `SetResource`. If `NewResource` is used to create the structure, the checks are redundant. Let's do without them for now.